Components/scripts/MapBase.gd

```diff
@@ -281,8 +284,58 @@ func wall(path):
 		changed_tiles[point + up] = true
 		changed_tiles[point + upup] = true
 		changed_tiles[point + downdown] = true
+	
+	return alreadyExistingWalls
+
+func possibleDoors(path):
+	var possibleDoorsInWalls = []
+	var possibleDoors = []
+	
+	if path.size() > 0:
+		var previousMovingHorizontal
+		
+		if (!is_corner_wall(path[0])):
+			possibleDoors.append(Vector2(path[0].x, path[0].y))
+		
+		if path.size() > 1:
+			for index in range(1, path.size()):
+				var point_a = path[index]
+				var point_b = path[index - 1]
+				var diff = point_b - point_a
+				var nextMoveHorizontal
+				
+				if diff[0] == 0 and diff[1] != 0:
+					nextMoveHorizontal = false
+				elif diff[0] != 0 and diff[1] == 0:
+					nextMoveHorizontal = true
+				else:
+					possibleDoorsInWalls.append(possibleDoors)
+					possibleDoors = []
+				
+				if previousMovingHorizontal == null && hasWallMoveChangedDirection(previousMovingHorizontal, nextMoveHorizontal) && possibleDoors.size() > 0:
+					possibleDoorsInWalls.append(possibleDoors)
+					possibleDoors = []
+				
+				previousMovingHorizontal = nextMoveHorizontal
+				
+				if (!is_corner_wall(point_a)):
+					possibleDoors.append(Vector2(point_a.x, point_a.y))
+	
+	possibleDoorsInWalls.append(possibleDoors)
+	
+	return possibleDoorsInWalls;
+
+func hasWallMoveChangedDirection(lastMoveHorizontal, nextMoveHorizontal):
+	return lastMoveHorizontal != nextMoveHorizontal
+
+func is_corner_wall(point):
+	var isVertical = is_door_or_wall(point + Vector2(0, 1)) || is_door_or_wall(point + Vector2(0, -1))
+	var isHorizontal = is_door_or_wall(point + Vector2(1, 0)) || is_door_or_wall(point + Vector2(-1, 0))
+	
+	return isVertical && isHorizontal
 
-		return multipleRoomWalls
+func is_door_or_wall(point):
+	return is_wall(tiles[point.y][point.x]) || is_door(point)
```

Comment:
I can't comment on is_door as it's not changed in this commit, but it seems really slow. I think we should maintain a dictionary of door tiles, which would make accessing it O(1).
e.g. we would
  - add a dict door_tiles = {}
  - Change add_door to set the tile to true. I'm pretty sure you can use Vector2's as a key as Vector2(1, 1) == Vector2(1, 1), unlike objects
  - Change is_door to just look up in the dictionary

As it stands it's O(N^2), and it checks all doors when a tile isn't a door or a wall(and most tiles aren't!)

Reply:
good idea